cpp/src/arrow/compute/exec/sink_node.cc

```diff
@@ -95,10 +95,15 @@ class SinkNode : public ExecNode {
 
   Future<> finished() override { return finished_; }
 
-  void InputReceived(ExecNode* input, ExecBatch batch) override {
+  void InputReceived(ExecNode* input, std::function<Result<ExecBatch>()> task) override {
     DCHECK_EQ(input, inputs_[0]);
 
-    bool did_push = producer_.Push(std::move(batch));
+    auto batch = task();
+    if (!batch.ok()) {
+      ErrorIfNotOk(batch.status());
+      return;
+    }
+    bool did_push = producer_.Push(batch.MoveValueUnsafe());
```

Comment:
This changes to submit task.

Reply:
SubmitTask is a method that is going to be part of the interface of the Scheduler, this part will need to changes in the next PRs related to the scheduler